cpp/src/arrow/stl_iterator.h

```diff
@@ -128,6 +131,148 @@ class ArrayIterator {
+  difference_type operator-(const ChunkedArrayIterator& other) const {
+    return index_ - other.index_;
+  }
+  ChunkedArrayIterator operator+(difference_type n) const {
+    return ChunkedArrayIterator(*chunked_array_, index_ + n);
+  }
+  ChunkedArrayIterator operator-(difference_type n) const {
+    return ChunkedArrayIterator(*chunked_array_, index_ - n);
+  }
+  friend inline ChunkedArrayIterator operator+(difference_type diff,
+                                               const ChunkedArrayIterator& other) {
+    return ChunkedArrayIterator(*other.chunked_array_, diff + other.index_);
+  }
+  friend inline ChunkedArrayIterator operator-(difference_type diff,
+                                               const ChunkedArrayIterator& other) {
+    return ChunkedArrayIterator(*other.chunked_array_, diff - other.index_);
+  }
+  ChunkedArrayIterator& operator+=(difference_type n) {
+    index_ += n;
+    auto chunk_location = GetChunkLocation(index_);
+    if (current_chunk_index_ == chunk_location.chunk_index) {
+      current_array_iterator_ -=
+          current_array_iterator_.index() - chunk_location.index_in_chunk;
+    } else {
+      current_array_iterator_ =
+          ArrayIterator<ArrayType>(*arrow::internal::checked_pointer_cast<ArrayType>(
+                                       chunked_array_->chunk(chunk_location.chunk_index)),
+                                   chunk_location.index_in_chunk);
+      current_chunk_index_ = chunk_location.chunk_index;
+    }
+    return *this;
+  }
+  ChunkedArrayIterator& operator-=(difference_type n) {
+    (*this) += -n;
+    return *this;
+  }
+
+  // Comparisons
+  bool operator==(const ChunkedArrayIterator& other) const {
+    return index_ == other.index_;
+  }
+  bool operator!=(const ChunkedArrayIterator& other) const {
+    return index_ != other.index_;
+  }
+  bool operator<(const ChunkedArrayIterator& other) const {
+    return index_ < other.index_;
+  }
+  bool operator>(const ChunkedArrayIterator& other) const {
+    return index_ > other.index_;
+  }
+  bool operator<=(const ChunkedArrayIterator& other) const {
+    return index_ <= other.index_;
+  }
+  bool operator>=(const ChunkedArrayIterator& other) const {
+    return index_ >= other.index_;
+  }
+
+ private:
+  arrow::internal::ChunkLocation GetChunkLocation(int64_t index) const {
+    return chunked_array_->chunk_resolver_.Resolve(index);
+  }
+
+  const ChunkedArray* chunked_array_;
+  int64_t index_;
+  int64_t current_chunk_index_;
+  ArrayIterator<ArrayType> current_array_iterator_;
+};
+
+template <typename Type, typename ArrayType = typename TypeTraits<Type>::ArrayType>
+ArrayIterator<ArrayType> Iterate(const Array& array) {
+  return stl::ArrayIterator<ArrayType>(&array);
+}
+
+template <typename Type, typename ArrayType = typename TypeTraits<Type>::ArrayType>
+ChunkedArrayIterator<ArrayType> Iterate(const ChunkedArray& chunked_array) {
+  return stl::ChunkedArrayIterator<ArrayType>(chunked_array);
+}
```

Comment:
Do we need these `Iterate` methods? One can simply create the corresponding iterator directly. Also, it is prone to error because it allows for the `Type` parameter to be different from the actual `Arrays` contained in the `chunked_array`.

On the other hand, if we are to keep factory functions, the convention is to prefix them with `Make`, so `MakeIterator` would be more clear, as `Iterate` is an action verb.

Reply:
I could add a condition for validate that and raise properly an error message, but I don't know if this is a good solution for that problem.

Reply:
Not ideal.

Reply:
@edponce Iterating is what this API is meant for. We have a separate `Iterator` concept in Arrow C++ that we don't want to mix up with this.